### ISSUE TITLE
Request to enable improved programatic error handling

### ISSUE THREAD

**Issue by jsimpson-gro** [awslabs/smithy-rs#2216]
When my application fails because it doesn't have the necessary credentials, I would like to have the application respond by renewing the credentials.

I have found it difficult to extract enough information from the error to determine it is a credentials related error. Essentially, I'm hoping to build a method that works like:

```
if is_authentication_error(e) {
    // Do stuff to renew credentials.
}
```

I have produced example errors by experimenting with three scenarios:
- no SSO login credentials present (`aws sso logout`)
- SSO credentials present but corrupted (use an editor to take a character out of the `accessToken` value)
- SSO credentials present and expired

PLEASE NOTE: These are experimentations of how to produce credential errors, they do not mean that I've been unable to correctly authenticate. I have a working application that correctly authenticates. I just want to respond to credential failures when those happen.

Printing out some errors, I see:

```
ConstructionFailure ConstructionFailure {
    source: CredentialsStageError {
        <stuff varies based on exact circumstances>
    },
}
```

According to the [documentation](https://docs.rs/aws-smithy-http/0.53.1/aws_smithy_http/result/enum.SdkError.html#variant.ConstructionFailure):

`The request failed during construction. It was not dispatched over the network.`

My interpretation is that `ConstructionFailure` is a very generic error. Just means an error before the request hits the network. That is not specific enough to decide that it is a credentials problem. So, I'm assuming that not all `ConstructionFailure`s are credential problems.

Reading the documentation for a `ConstructionFailure` I don't see how to navigate any deeper in the error hierarchy.

I've searched in the [aws-smithy-http docs](https://docs.rs/aws-smithy-http/0.53.1/aws_smithy_http/result/enum.SdkError.html?search=Credentials) and the [aws_sdk_s3](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/types/enum.SdkError.html?search=Credentials) and I don't see a `CredentialsStageError` in the docs.

Additional note: When I corrupt the SSO token in the `~/.aws/sso/cache/<file-name>.json` by editing the `accessToken` value, the error produced appears to have accessed the network. When the whole structure is printed out, there is a `Response` section in the printout, with a 401 status, and RequestId headers. This leads me to believe that a `ConstructionFailure` may involve network activity, counter to the documentation.

**Comment by Velfi** [comment]
Have you tried running your experiments with [logging enabled](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/logging.html#tracing)? That could make things clearer for you.

**Comment by ysaito1001** [comment]
> This leads me to believe that a ConstructionFailure may involve network activity

Correct, if a credentials provider needs to fetch credentials via HTTP(S) as part of constructing a request to  the target service (s3 in your example).

> counter to the documentation.

I believe the documentation is correct in that the request itself was not dispatched to S3.

**Comment by jsimpson-gro** [comment]
> What are you trying to accomplish in your code?
> 
> > When my application fails because it doesn't have the necessary credentials
> 
> Are you aware that our default credentials providers already handle this? What's the use of implementing your own handler for this?

The default credentials provider will restart my application if the AWS credentials from the environment variables don't work?

**Comment by ysaito1001** [comment]
The default credentials provider forms a chain of providers ([doc](https://docs.rs/aws-config/0.53.0/aws_config/default_provider/credentials/struct.DefaultCredentialsChain.html)).

If credentials are not found from the environment variable, the next credentials provider in the chain is consulted, and so on, until either credentials are returned from one of the provides or no credentials are loaded from the entire chain. If it's the latter, a [CredentialsError](https://docs.rs/aws-credential-types/0.53.0/aws_credential_types/provider/error/enum.CredentialsError.html) is returned.

Does your use case involve something `DefaultCredentialsChain` cannot handle?
